xarray/core/computation.py

```diff
@@ -1142,6 +1142,42 @@ def earth_mover_distance(first_samples,
         return apply_array_ufunc(func, *args, dask=dask)
 
 
+def apply_to_dataset(func, obj, *args, **kwargs):
+    """apply a function expecting a Dataset to a xarray object
+
+    Parameters
+    ----------
+    func : callable
+        A function expecting a Dataset as its first parameter.
+    obj : DataArray or Dataset
+        The dataset to apply ``func`` to. If a ``DataArray``, convert it to a single
+        variable ``Dataset`` first.
+    *args, **kwargs
+        Additional arguments to ``func``
+
+    Returns
+    -------
+    DataArray or Dataset
+        The result of ``func(obj, *args, **kwargs)`` with the same type as ``obj``.
+
+    Notes
+    -----
+    If a ``DataArray``, result will have the same name as ``obj`` but the single data
+    variable in the temporary ``Dataset`` will always have a generic name.
```

Comment:
Is this as simple as "``DataArray``s will retain their name"? If so, maybe we don't need any notes? (very possible I'm missing some of the complexity, as ever)

Reply:
no, I might not have explained that correctly. The temporary `Dataset` generates always has a `<this-array>` variable, but the original name will be restored by `_from_temp_dataset`.

Reply:
Does the name of the variable of the temporary dataset matter to the user though? To what extent is that just an implementation detail?

Reply:
the function will see the `Dataset` so it might be important to keep the note. For example, [this](https://github.com/xarray-contrib/pint-xarray/blob/038ac601c966fcb6d15e400d67e88ecaf87339b5/pint_xarray/conversion.py#L181-L192) would need to change the name in the `units` dict from `None` to `<this-array>` to work correctly.

Reply:
in `core/parallel.py` there's a `dataarray_to_dataset` (and inverse `dataset_to_dataarray`) function that preserves name if possible. I think name preservation is a good thing for a user-facing function.